src/main/java/org/springframework/data/redis/cache/RedisCache.java

```diff
@@ -84,190 +202,395 @@ public Object getNativeCache() {
+			} finally {
+				connection.del(metadata.getCacheLockKey());
+			}
+		}
+
+		public abstract T doInLock(RedisConnection connection);
+
 	}
 
-	public void clear() {
-		// need to del each key individually
-		template.execute(new RedisCallback<Object>() {
-			public Object doInRedis(RedisConnection connection) throws DataAccessException {
-				// another clear is on-going
-				if (connection.exists(cacheLockName)) {
-					return null;
-				}
+	/**
+	 * @author Christoph Strobl
+	 * @since 1.5
+	 */
+	static class RedisCacheCleanByKeysCallback extends LockingRedisCacheCallback<Void> {
 
-				try {
-					connection.set(cacheLockName, cacheLockName);
+		private static final int PAGE_SIZE = 128;
+		private final RedisCacheMetadata metadata;
 
-					int offset = 0;
-					boolean finished = false;
+		RedisCacheCleanByKeysCallback(RedisCacheMetadata metadata) {
+			super(metadata);
+			this.metadata = metadata;
+		}
 
-					do {
-						// need to paginate the keys
-						Set<byte[]> keys = connection.zRange(setName, (offset) * PAGE_SIZE, (offset + 1) * PAGE_SIZE - 1);
-						finished = keys.size() < PAGE_SIZE;
-						offset++;
-						if (!keys.isEmpty()) {
-							connection.del(keys.toArray(new byte[keys.size()][]));
-						}
-					} while (!finished);
+		@Override
+		public Void doInLock(RedisConnection connection) {
 
-					connection.del(setName);
-					return null;
+			int offset = 0;
+			boolean finished = false;
 
-				} finally {
-					connection.del(cacheLockName);
+			do {
+				// need to paginate the keys
+				Set<byte[]> keys = connection.zRange(metadata.getSetOfKnownKeysKey(), (offset) * PAGE_SIZE, (offset + 1)
+						* PAGE_SIZE - 1);
+				finished = keys.size() < PAGE_SIZE;
+				offset++;
+				if (!keys.isEmpty()) {
+					connection.del(keys.toArray(new byte[keys.size()][]));
 				}
-			}
-		}, true);
+			} while (!finished);
+
+			connection.del(metadata.getSetOfKnownKeysKey());
+			return null;
+		}
 	}
 
-	private byte[] computeKey(Object key) {
+	/**
+	 * @author Christoph Strobl
+	 * @since 1.5
+	 */
+	static class RedisCacheCleanByPrefixCallback extends LockingRedisCacheCallback<Void> {
 
-		byte[] keyBytes = convertToBytesIfNecessary(template.getKeySerializer(), key);
+		private static final String REMOVE_KEYS_BY_PATTERN_LUA = "local keys = redis.call('KEYS', ARGV[1]); local keysCount = table.getn(keys); if(keysCount > 0) then for _, key in ipairs(keys) do redis.call('del', key); end; end; return keysCount;";
+		private final RedisCacheMetadata metadata;
 
-		if (prefix == null || prefix.length == 0) {
-			return keyBytes;
+		public RedisCacheCleanByPrefixCallback(RedisCacheMetadata metadata) {
+			super(metadata);
+			this.metadata = metadata;
 		}
 
-		byte[] result = Arrays.copyOf(prefix, prefix.length + keyBytes.length);
-		System.arraycopy(keyBytes, 0, result, prefix.length, keyBytes.length);
+		@Override
+		public Void doInLock(RedisConnection connection) throws DataAccessException {
+
+			byte[] wildCard = "*".getBytes();
```

Comment:
Use `StringRedisSerializer` to ensure property serialization to binary representation.